### ISSUE TITLE
atomiccorr: Internal Error: Attempting to assign to Frame with external memory

### ISSUE THREAD

**Issue by hainm** [Amber-MD/cpptraj#172]
This is not a fatal error, I still got identical output by either out-of-core trajectory or in-memory traj (with fast frame iteration by setting external memory view).

```python
In [4]: traj 
Out[4]: 
pytraj.TrajectoryIterator, 10 frames: 
Size: 0.001183 (GB)
<Topology: 5293 atoms, 1704 residues, 1692 mols, PBC with box type = ortho>
           

In [5]: # load two frames to memory and use fast Frame iter

In [6]: t0 = traj[[0, 3]]

In [7]: pt.atomiccorr(t0, '@CA')
Internal Error: Attempting to assign to Frame with external memory.
Out[7]: 
array([[ 0.        ,  0.66436392,  0.08319134, ...,  0.62662166,
         0.99535167,  0.84034395],
       [ 0.66436392,  0.        ,  0.79999709, ...,  0.89398563,
         0.71788037,  0.16386618],
       [ 0.08319134,  0.79999709,  0.        , ...,  0.69588429,
         0.15919216, -0.45399991],
       ..., 
       [ 0.62662166,  0.89398563,  0.69588429, ...,  0.        ,
         0.69864297,  0.24450834],
       [ 0.99535167,  0.71788037,  0.15919216, ...,  0.69864297,
         0.        ,  0.80383575],
       [ 0.84034395,  0.16386618, -0.45399991, ...,  0.24450834,
         0.80383575,  0.        ]], dtype=float32)

In [8]: # use out-of-core traj, loada single Frame at a time

In [9]: pt.atomiccorr(traj, '@CA', frame_indices=[0, 3])
Out[9]: 
array([[ 0.        ,  0.66436392,  0.08319134, ...,  0.62662166,
         0.99535167,  0.84034395],
       [ 0.66436392,  0.        ,  0.79999709, ...,  0.89398563,
         0.71788037,  0.16386618],
       [ 0.08319134,  0.79999709,  0.        , ...,  0.69588429,
         0.15919216, -0.45399991],
       ..., 
       [ 0.62662166,  0.89398563,  0.69588429, ...,  0.        ,
         0.69864297,  0.24450834],
       [ 0.99535167,  0.71788037,  0.15919216, ...,  0.69864297,
         0.        ,  0.80383575],
       [ 0.84034395,  0.16386618, -0.45399991, ...,  0.24450834,
         0.80383575,  0.        ]], dtype=float32)

```

It's likely that the error warning is due to the swapping 'refframe_ = frm.Frm()' https://github.com/Amber-MD/cpptraj/blob/master/src/Action_AtomicCorr.cpp#L143

So how can I avoid this? (making a copy of Frame so we release the 'view'?) https://github.com/Amber-MD/cpptraj/blob/master/src/Frame.cpp#L216-L222

**Comment by hainm** [comment]
Since I don't see the the reference Frame section, so I am assuming that https://github.com/Amber-MD/cpptraj/blob/master/src/Action_AtomicCorr.cpp is in-progress code?

**Comment by drroe** [comment]
> Since I don't see the the reference Frame section, so I am assuming that https://github.com/Amber-MD/cpptraj/blob/master/src/Action_AtomicCorr.cpp is in-progress code?

Not sure what you are talking about - `atomiccorr` has been a documented Action in cpptraj for a while now (see current Amber 15 manual section 29.9.3).

The problem here is the Frame assignment operator currently doesn't have code implemented to go from external memory to internal (see Frame.cpp:217). The code just needs to be written.

**Comment by hainm** [comment]
> The problem here is the Frame assignment operator currently doesn't have code implemented to go from external memory to internal (see Frame.cpp:217). The code just needs to be written.

got it.

**Comment by hainm** [comment]
I just want to update my 1st post after varying number of frames

"This is **not** a fatal error" should be "This is a fatal error" 

The example I showed work well with slicing only two frames (frame_indices=[0, 3]) but the output does not match (between in and out of memory) with more than 2 frames.

Out-of memory calculation

```python
In [3]: pt.atomiccorr(traj(0, 8, 2), '@CA')
Out[3]: 
array([[ 0.        ,  0.78620476,  0.57646966, ...,  0.70500332,
         0.71050465,  0.65380627],
       [ 0.78620476,  0.        ,  0.9323197 , ...,  0.66661882,
         0.42925909,  0.30514067],
       [ 0.57646966,  0.9323197 ,  0.        , ...,  0.49003574,
         0.19772419,  0.05126373],
       ..., 
       [ 0.70500332,  0.66661882,  0.49003574, ...,  0.        ,
         0.89304864,  0.68868572],
       [ 0.71050465,  0.42925909,  0.19772419, ...,  0.89304864,
         0.        ,  0.8909083 ],
       [ 0.65380627,  0.30514067,  0.05126373, ...,  0.68868572,
         0.8909083 ,  0.        ]], dtype=float32)
```

In memory calculation (load all frames to memory)
```
In [4]: pt.atomiccorr(traj[0:8:2], '@CA')
Internal Error: Attempting to assign to Frame with external memory.
Internal Error: Attempting to assign to Frame with external memory.
Internal Error: Attempting to assign to Frame with external memory.
Out[4]: 
array([[ 0.        ,  0.70406234,  0.62737095, ...,  0.92045403,
         0.90662849,  0.68665415],
       [ 0.70406234,  0.        ,  0.92030787, ...,  0.78862607,
         0.57772225,  0.09668256],
       [ 0.62737095,  0.92030787,  0.        , ...,  0.74814546,
         0.46166068,  0.02892438],
       ..., 
       [ 0.92045403,  0.78862607,  0.74814546, ...,  0.        ,
         0.87096745,  0.49775332],
       [ 0.90662849,  0.57772225,  0.46166068, ...,  0.87096745,
         0.        ,  0.76173103],
       [ 0.68665415,  0.09668256,  0.02892438, ...,  0.49775332,
         0.76173103,  0.        ]], dtype=float32)
```

**Comment by drroe** [comment]
Should be fixed by #176

**Comment by hainm** [comment]
cool. confirmed

```python
In [1]: pt.atomiccorr(traj(0, 8, 2), '@CA')
Out[1]: 
array([[ 0.        ,  0.78620476,  0.57646966, ...,  0.70500332,
         0.71050465,  0.65380627],
       [ 0.78620476,  0.        ,  0.9323197 , ...,  0.66661882,
         0.42925909,  0.30514067],
       [ 0.57646966,  0.9323197 ,  0.        , ...,  0.49003574,
         0.19772419,  0.05126373],
       ..., 
       [ 0.70500332,  0.66661882,  0.49003574, ...,  0.        ,
         0.89304864,  0.68868572],
       [ 0.71050465,  0.42925909,  0.19772419, ...,  0.89304864,
         0.        ,  0.8909083 ],
       [ 0.65380627,  0.30514067,  0.05126373, ...,  0.68868572,
         0.8909083 ,  0.        ]], dtype=float32)

In [2]: pt.atomiccorr(traj[0:8:2], '@CA')
Out[2]: 
array([[ 0.        ,  0.78620476,  0.57646966, ...,  0.70500332,
         0.71050465,  0.65380627],
       [ 0.78620476,  0.        ,  0.9323197 , ...,  0.66661882,
         0.42925909,  0.30514067],
       [ 0.57646966,  0.9323197 ,  0.        , ...,  0.49003574,
         0.19772419,  0.05126373],
       ..., 
       [ 0.70500332,  0.66661882,  0.49003574, ...,  0.        ,
         0.89304864,  0.68868572],
       [ 0.71050465,  0.42925909,  0.19772419, ...,  0.89304864,
         0.        ,  0.8909083 ],
       [ 0.65380627,  0.30514067,  0.05126373, ...,  0.68868572,
         0.8909083 ,  0.        ]], dtype=float32)
```
